extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -2689,20 +2711,15 @@ private void startStages() throws IOException, InterruptedException
           }
 
           final StageId shuffleStageId = new StageId(queryDef.getQueryId(), shuffleStageNumber);
-          final boolean isTimeBucketed = isTimeBucketedIngestion(task.getQuerySpec());
-          final ClusterByPartitions partitionBoundaries =
-              queryKernel.getResultPartitionBoundariesForStage(shuffleStageId);
+          final Boolean isShuffleStageOutputEmpty = queryKernel.isStageOutputEmpty(shuffleStageId);
 
-          // We require some data to be inserted in case it is partitioned by anything other than all and we are
-          // inserting everything into a single bucket. This can be handled more gracefully instead of throwing an exception
-          // Note: This can also be the case when we have limit queries but validation in Broker SQL layer prevents such
-          // queries
-          if (isTimeBucketed && partitionBoundaries.equals(ClusterByPartitions.oneUniversalPartition())) {
+          if (!isFailOnEmptyInsertEnabled && Boolean.TRUE.equals(isShuffleStageOutputEmpty)) {
```

Comment:
I think this might end up being unknown in some situations where we're using `PARTITIONED BY ALL` or `LIMIT`. I don't see tests for those, so please add them, and then let's consider what that tells us.

I suspect that there may be some cases where the "is output empty" ends up being unknown when there is in fact no data, meaning the `failOnEmptyInsert: true` won't always behave like we expect it to. (Sometimes the empty insert will go through even with `failOnEmptyInsert: true`)

Reply:
Added unit tests for `ALL` grain and `LIMIT` queries with:
- `failOnEmptyInsert: true`  in `MSQFaultTest`
- `failOnEmptyInsert: false` in `MSQInsertTest` and `MSQReplaceTest`

The output stage is empty (`true`) for all these cases. Is there a specific unknown case you can think of?